Mathlib/Analysis/Complex/UpperHalfPlane/MoebiusAction.lean

```diff
@@ -41,113 +40,174 @@ scoped notation:1024 "↑ₘ[" R "]" A:1024 =>
+    Complex.ofReal_mul, num]
+  ring
+
+lemma moebius_im (g : GL (Fin 2) ℝ) (z : ℂ) :
+    (num g z / denom g z).im = g.det.val * z.im / Complex.normSq (denom g z) := by
+  simp only [num, denom, Complex.div_im, Complex.add_im, Complex.mul_im, Complex.ofReal_re,
+    Complex.ofReal_im, zero_mul, add_zero, Complex.add_re, Complex.mul_re, sub_zero, ← sub_div,
+    GeneralLinearGroup.val_det_apply, g.1.det_fin_two]
   ring
 
+/-- Automorphism of `ℂ`: the identity if `0 < det g` and conjugation otherwise. -/
+def σ (g : GL (Fin 2) ℝ) : ℂ →+* ℂ := if 0 < g.det.val then RingHom.id ℂ else starRingEnd ℂ
+
+lemma σ_ofReal (g : GL (Fin 2) ℝ) (y : ℝ) : σ g y = y := by
+  simp only [σ]
+  split_ifs <;> simp
+
+lemma σ_num (x y : GL (Fin 2) ℝ) (z : ℂ) : σ x (num y z) = num y (σ x z) := by
+  simp only [num, σ_ofReal, map_add, map_mul]
+
+lemma σ_denom (x y : GL (Fin 2) ℝ) (z : ℂ) : σ x (denom y z) = denom y (σ x z) := by
+  simp only [denom, σ_ofReal, map_add, map_mul]
+
+lemma σ_sq (g : GL (Fin 2) ℝ) (x : ℂ) : σ g (σ g x) = x := by
+  simp only [σ]
+  split_ifs <;> simp
+
+lemma σ_im_ne_zero {g x} : (σ g x).im ≠ 0 ↔ x.im ≠ 0 := by
+  simp only [σ]
+  split_ifs <;> simp
+
+lemma σ_mul (g g' : GL (Fin 2) ℝ) (x : ℂ) : σ (g * g') x = σ g (σ g' x) := by
+  simp only [σ, map_mul, Units.val_mul]
+  rcases g.det_ne_zero.lt_or_lt with (h | h) <;>
+  rcases g'.det_ne_zero.lt_or_lt with (h' | h')
+  · simp [mul_pos_of_neg_of_neg h h', h.not_lt, h'.not_lt]
+  · simp [(mul_neg_of_neg_of_pos h h').not_lt, h.not_lt, h']
+  · simp [(mul_neg_of_pos_of_neg h h').not_lt, h, h'.not_lt]
+  · simp [mul_pos h h', h, h']
+
+lemma σ_mul_comm (g g' : GL (Fin 2) ℝ) (x : ℂ) : σ g (σ g' x) = σ g' (σ g x) := by
+  simp only [σ]
+  split_ifs <;> simp
+
+/-- Fractional linear transformation, also known as the Moebius transformation -/
+def smulAux' (g : GL (Fin 2) ℝ) (z : ℂ) : ℂ := σ g (num g z / denom g z)
+
+lemma smulAux'_im (g : GL (Fin 2) ℝ) (z : ℂ) :
+    (smulAux' g z).im = |g.det.val| * z.im / Complex.normSq (denom g z) := by
+  simp only [smulAux', σ]
+  split_ifs with h <;>
+  [rw [abs_of_pos h]; rw [abs_of_nonpos (not_lt.mp h)]] <;>
+  simpa only [starRingAut_apply, Complex.star_def, Complex.conj_im,
+    neg_mul, neg_div, neg_inj] using moebius_im g z
+
 /-- Fractional linear transformation, also known as the Moebius transformation -/
-def smulAux (g : GL(2, ℝ)⁺) (z : ℍ) : ℍ :=
+def smulAux (g : GL (Fin 2) ℝ) (z : ℍ) : ℍ :=
   mk (smulAux' g z) <| by
     rw [smulAux'_im]
-    convert mul_pos ((mem_glpos _).1 g.prop)
-        (div_pos z.im_pos (Complex.normSq_pos.mpr (denom_ne_zero g z))) using 1
-    simp only [GeneralLinearGroup.val_det_apply]
-    ring
+    exact div_pos (mul_pos (abs_pos.mpr g.det.ne_zero) z.im_pos) (normSq_denom_pos _ z.im_ne_zero)
 
-theorem denom_cocycle (x y : GL(2, ℝ)⁺) (z : ℍ) :
-    denom (x * y) z = denom x (smulAux y z) * denom y z := by
+lemma denom_cocycle' (x y : GL (Fin 2) ℝ) (z : ℍ) :
```

Comment:
having  `x,y` here (and below) for elements of GL and `g` elsewhere is a bit confusing. Maybe call them `f,g` or something?